extensions/azure/storage/AzureDataLakeStorageClient.cpp

```diff
@@ -0,0 +1,55 @@
+/**
+ * @file AzureDataLakeStorageClient.cpp
+ * AzureDataLakeStorageClient class implementation
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "AzureDataLakeStorageClient.h"
+
+namespace org::apache::nifi::minifi::azure::storage {
+
+void AzureDataLakeStorageClient::resetClientIfNeeded(const std::string& connection_string, const std::string& file_system_name) {
+  if (client_ == nullptr || connection_string_ != connection_string || file_system_name_ != file_system_name) {
+    client_ = std::make_unique<Azure::Storage::Files::DataLake::DataLakeFileSystemClient>(
+      Azure::Storage::Files::DataLake::DataLakeFileSystemClient::CreateFromConnectionString(connection_string, file_system_name));
+    file_system_name_ = file_system_name;
+    connection_string_ = connection_string;
+  }
+}
+
+Azure::Storage::Files::DataLake::DataLakeFileClient AzureDataLakeStorageClient::getFileClient(const PutAzureDataLakeStorageParameters& params) {
+  resetClientIfNeeded(params.connection_string, params.file_system_name);
+  auto directory_client = client_->GetDirectoryClient(params.directory_name);
+  if (!params.directory_name.empty()) {
+    directory_client.CreateIfNotExists();
+  }
+  return directory_client.GetFileClient(params.filename);
+}
+
+bool AzureDataLakeStorageClient::createFile(const PutAzureDataLakeStorageParameters& params) {
+  auto file_client = getFileClient(params);
+  auto response = file_client.CreateIfNotExists();
+  return response.Value.Created;
```

Comment:
Is it possible to detect why the file creation failed?  I would expect three different return values here: Success, File Already Exists, and Other Error.  In the last case we don't want to ignore the error, whatever the Conflict Resolution Strategy is.

Reply:
It is done in the AzureDataLakeStorage class. If an error occurs it throws an exception from the SDK, but if the file already exists it does not throw any errors, only sets the `response.Value.Created` value false. In AzureDataLakeStorage::uploadFile we return these 3 possible values represented in the UploadResultCode enum. I wanted to minimize the functionality of the AzureDataLakeStorageClient as it is mocked in the tests, so it should mostly be a wrapper of the Azure SDK client.